### DISCUSSION TITLE
RDF of NaCl in water

### DISCUSSION THREAD

**Discussion by horlust** [MDAnalysis/mdanalysis#4437] [Q&A] [answered]
Hi everyone,

I have a simulation containing 884 water molecules, 1 Na+ atom and 1 Cl- atom. I am trying to compute the RDF of Na+ and Cl- in water and also the coordination number.

At first I did:

```
from MDAnalysis.analysis.rdf import InterRDF

# select atoms
na_selection = u.select_atoms("name NA")
water_selection = u.select_atoms("resname SOL")

# define the range for RDF calculation
r_min = 0.0  # minimum distance (in Angstroms)
r_max = 10.0  # maximum distance (in Angstroms)
nbins = 100  # number of bins for RDF

# calculate RDF
rdf_nacl = InterRDF(na_selection, 
                        water_selection, 
                        nbins=nbins, 
                        range=(r_min, r_max), 
                        periodic=True  # account for periodic boundary conditions
                        )
rdf_nacl.run()
```

And using a function I calculated the coordination number of Na+ in this case, which was close to 6. However, if you look at my code you will see that I've calculated RDF with respect to the water molecules (water_selection is  "resname SOL"). Is the RDF function on MDAnalysis capable of performing this calculation for a whole molecule, instead of just for atoms? I am asking because pretty much all RDF calculations that I've seen consider the oxygen atoms of the water molecule instead of the whole water molecule. So I also did this:

```
# select atoms
na_selection = u.select_atoms("name NA")
water_selection = u.select_atoms("resname SOL and name OW")

# define the range for RDF calculation
r_min = 0.0  # minimum distance (in Angstroms)
r_max = 10.0  # maximum distance (in Angstroms)
nbins = 100  # number of bins for RDF

# calculate RDF
rdf_nacl = InterRDF(na_selection, 
                        water_selection, 
                        nbins=nbins, 
                        range=(r_min, r_max), 
                        periodic=True  # account for periodic boundary conditions
                        )
rdf_nacl.run()
```

In this case, I consider only the oxygen atoms of the water molecules around Na+.

As I'm having two very different results (images attached), I'd like to understand the handling of the RDF function. Which way of calculating the RDF and the coordination number makes more sense in the context of MDAnalysis?

Thank you in advance!
Best,
Maria

![rdf_na_water_1](https://github.com/MDAnalysis/mdanalysis/assets/50314998/5e08fa7e-9512-4d16-ad61-514f701303df)
![rdf_na_water_2](https://github.com/MDAnalysis/mdanalysis/assets/50314998/04d9ac1f-f91b-42c7-941c-3499431fe682)

**Comment by hmacdope** [comment]
Hi! 

Could you clarify which graph is which experiment? 

I am assuming that the 5.2 coordination number is the one using water oxygens. From my knowledge use of the oxygens as the reference is standard and gives a number that makes a lot of sense,  octahedral with ~ 6 * 1 coordinating atoms  (likely has 0.8 fractional Cl- coordination). 

Using all the solvent molecule as a whole gives a number that also makes sense (~6 * 3 atoms per molecule), but is not the standard way to talk about water coordination, as the oxygen is normally the coordinating atom. 

You can see from the data tables in a paper parametrising ions such as this one: https://pubs.acs.org/doi/10.1021/jp505875v  that coordination numbers are normally reported as directly chelating atoms only.

If you are looking at data of this kind, I would reccomend our solvation subpackage `solvation-analysis`: https://github.com/MDAnalysis/solvation-analysis

Hope this helps. 


Hugo

**Reply by horlust** [reply]
Hi Hugo!

Thank you for your answer :-)

In fact, for coordination number = 5.2 I used the whole molecule of water. For coordination number = 17 I used the water oxygens. I also used Solvation Analysis and I got this for the same NaCl simulation: 

![na_solvation_analysis](https://github.com/MDAnalysis/mdanalysis/assets/50314998/7acc4c30-2439-443e-b038-2c5d32585b9a)

I think the Solvation Analysis output is closer to the calculation considering the whole water molecule. That's why I'm a bit confused. Maybe my coordination number calculation from RDF data is wrong too? From what I've read it should be close to 6 for Na+, but with the MDAnalysis RDF calculation I get 17. I did something like:

```
`# numerical density calculation
def calc_density(u):
    pbc = u.atoms.bbox(pbc=True)
    x_l = (pbc[1][0] - pbc[0][0])
    y_l = (pbc[1][1] - pbc[0][1])
    z_l = (pbc[1][2] - pbc[0][2])
    volume = x_l*y_l*z_l
    n_part = len(u.atoms)
    density = n_part/volume
    return density

# coordination number calculation
def calc_coord_number(r_polyol, g_r_polyol, r_min, r_max, rho):
    # interpolate g(r) function
    g_interpolated = np.interp

    # define the function to integrate
    def integrand(r):
        return g_interpolated(r, r_polyol, g_r_polyol) * r**2

    # perform numerical integration
    result, _ = integrate.quad(integrand, r_min, r_max)

    # calculate CN(r)
    CN_r = 4 * np.pi * rho * result

    return CN_r

  # find the first local minimum in g(r)
  minima_index = np.where(np.r_[True, g_r_nacl[1:] < g_r_nacl[:-1]] & np.r_[g_r_nacl[:-1] < g_r_nacl[1:], True])[0][0]
  
  # define the integration limits
  r_min_integrate = 0
  r_max_integrate = r_nacl[minima_index]
  
  CN_r = calc_coord_number(r_nacl, g_r_nacl, r_min_integrate, r_max_integrate, calc_density(u))`
```

Thanks again, your answer have been very helpful.
Maria

**Reply by rodpollet** [reply, marked as answer]
Hi Maria,

I checked your code and found two mistakes.

First, I think you should not use bbox but u.trajectory.ts.dimensions, namely:
x_l = u.trajectory.ts.dimensions[0]
y_l = u.trajectory.ts.dimensions[1]
z_l = u.trajectory.ts.dimensions[2]
This should give a better estimate of the volume.

Second, I think you should not set n_part to the total number of atoms in your universe but only the number of atoms in the second species, namey water oxygen atoms only. Again, this will correct the density.

With these corrections, I think you should obtain the correct result. Do you confirm?

Hope this helps,
Rodolphe

**Reply by horlust** [reply]
Hi Rodolphe,

I modified the code according to your corrections and I got a coordination number of 5.65 for Na+ and 6.84 for Cl-, for the ion regarding the oxygen atoms. I think it is correct now. Thanks!

**Reply by hmacdope** [reply]
Amazing! Thanks @rodpollet!

**Reply by orbeckst** [reply]
Now that we know that it's the volume, I'd just use the built-in unit cell volume calculation, which will be correct for any box shape:
```python
volume = u.trajectory.ts.volume
```

**Reply by horlust** [reply]
@rodpollet, I kept thinking about your answer and I don't understand why n_part should be the number of oxygen atoms and not the number of atoms "tout court". Can you explain? :)

**Reply by rodpollet** [reply]
@horlust Because you have to compute the density of oxygen atoms at distance r from atoms that belong to your other species.

By the way I wonder if there is an error in the first formula of [this page of the user guide](https://userguide.mdanalysis.org/stable/examples/analysis/structure/average_rdf.html). I think the total volume V is missing in the numerator (which would agree with the _conclude function of the InterRDF class in the rdf.py source file in the library). @horlust , if you multiply g<sub>ab</sub>(r) by N<sub>b</sub>/V, you will see that the density of particles b at a distance r around particles a is proportional to 1/N<sub>a</sub> times the double sum (i.e., an average count).

**Reply by orbeckst** [reply]
The $\delta$-function (in 3d) has units of 1/Volume and the density of $N$ particles is formally

$$
n(\mathbf{x}) = \frac{1}{N}\sum_{\alpha=1}^N \delta(\mathbf{x} - \mathbf{x}\_\alpha) = \langle \delta(\mathbf{x} - \mathbf{x}\_\alpha ) \rangle
$$

(see e.g. Chaikin & Lubensky, Principles of Condensed Matter Physics, Eq 2.3.1).

The radial distribution function is a 2-particle density. In the User Guide, the "division by volume" is implicit in the use of the delta-function.

**Reply by rodpollet** [reply]
Yes but the following lines in rdf.py seem to indicate that 1/Volume is used instead of 1/(Volume^2):
N = nA * nB
norm *= N / box_vol
IMHO this is why we need to multiply by Nb/V to get the coordination number. Am I wrong?

**Reply by orbeckst** [reply]
From the code I see that we calculate
```
g(r) = {[counts(r) / DeltaV(r)] / [nb/V]} / na
```
i.e., "[density of b-particles in shell at r] divided by [average b density] and averaged over all na reference points" (and we also average over all frames). `N=na*nb` is used in the code as a shorthand for "all pairs". This makes the RDF have the proper units of 1. The code all looks correct to me.

Looking at the equation for $g\_{ab}$ I am actually missing a a factor of $V$ so I think it should read:

$$
g_{ab}(r) = \frac{V}{N\_{a} N\_{b}} \sum_{i=1}^{N\_a} \sum_{j=1}^{N\_b} \langle \delta(|\mathbf{r}\_i - \mathbf{r}\_j| - r) \rangle
$$

i.e.,

$$
g_{ab}(r) = \frac{1}{N\_{a}} \frac{1}{N\_{b}/V} \sum_{i=1}^{N\_a} \sum_{j=1}^{N\_b} \langle \delta(|\mathbf{r}\_i - \mathbf{r}\_j| - r) \rangle
$$

The other equations should then make sense.

## code 
https://github.com/MDAnalysis/mdanalysis/blob/73acc9bf9631c135dfd40322e121f379abfce198/package/MDAnalysis/analysis/rdf.py#L294-L298
https://github.com/MDAnalysis/mdanalysis/blob/73acc9bf9631c135dfd40322e121f379abfce198/package/MDAnalysis/analysis/rdf.py#L304
https://github.com/MDAnalysis/mdanalysis/blob/73acc9bf9631c135dfd40322e121f379abfce198/package/MDAnalysis/analysis/rdf.py#L314
https://github.com/MDAnalysis/mdanalysis/blob/73acc9bf9631c135dfd40322e121f379abfce198/package/MDAnalysis/analysis/rdf.py#L316

**Reply by orbeckst** [reply]
I am correcting the RDF equation in
* https://github.com/MDAnalysis/mdanalysis/pull/4591 (for the API docs)
* https://github.com/MDAnalysis/UserGuide/pull/367 (for the User Guide)

**Comment by orbeckst** [comment]
For RDF calculations if ions in water I would choose ion-oxygen. The MDA [`InterRDF()`](https://docs.mdanalysis.org/stable/documentation_pages/analysis/rdf.html#MDAnalysis.analysis.rdf.InterRDF) class will compute RDFs for atom-atom distances; therefore in your first example you get a superposition of the Na-O and the NA-H RDFs. There is not option for atom-molecule or molecule-molecule evaluations at the moment.

**Comment by hmacdope** [comment]
@horlust is your question answered? If so can you mark one of the answers as such.
